rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+
+This option suggests making the cache a hidden feature without providing any
+control over its capacity.
+
+There are two approaches to defining the cache capacity:
+1. Define a reasonable constant
+2. Tie the cache capacity to the primitive cache capacity
+
+The first approach is overly rigid and it is hard to define a reasonable
+capacity that would work for the majority of workloads therefore we will consider
+the second one.
+
+Pros:
+* Provides sufficient control to benefit from the cache and to maintain memory
+consumption
+* Doesn't introduce yet another *cache capacity* definition that will simplify using the cache
+
+Cons:
+* There is no guarantee that all kernels in the kernel cache are used by all the
+primitives in the primitive cache. It might happen those primitives in the
+primitive cache has unique kernels and the kernel cache may have unique kernels
+too. This means that memory consumption can be higher that it is now
+
+##### Option 2: Provide API to Control Kernel Cache Capacity
+
+This option suggests adding an API to control kernel cache capacity similar to
+the one for the primitive cache.
+
+Pros:
+* Provides more control over the kernel cache and therefore more flexibility to
+the users. Those users who heavily rely on dynamic shapes can balance between
+primitive cache capacity and the kernel cache capacity
+* User can fully control memory consumption
+
+Cons:
+* Introducing yet another *cache capacity* definition that should be specified.
+Also, we will have to explain to the users the difference between kernel cache and primitive cache capacities as well as how they interoperate with each other
+
+###### C API
+
+```c
+dnnl_status_t dnnl_get_kernel_cache_capacity(int *capacity);
+dnnl_status_t dnnl_set_kernel_cache_capacity(int capacity);
+```
+
+###### C++ API
+
+```cpp
+int get_primitive_cache_capacity();
+void set_primitive_cache_capacity(int capacity);
+```
+
+###### Environment Variable
+
+`ONEDNN_PRIMITIVE_CACHE_CAPACITY=<number>`, where `number` is the kernel cache
+capacity. The default value is 1024. When `number` is 0 the kernel cache is
+disabled.
+
+###### Build Time Control
+
+The kernel cache can be enabled or disabled at the build time using a CMake option
+`ONEDNN_ENABLE_KERNEL_CACHE`.
+
+##### Options Comparison
+
+Both options assume that there will be a reasonable default capacity for the
+kernel cache. Also, both options allow users to control the caches capacity
+therefore, the only difference between the options is in the flexibility.
+The second option is in fact the first one with advanced controlling
+capabilities.
+
+Given that the first option will address the problem with dynamic shapes
+related creation overhead the proposal is to start with the it and add the
+API from the option two if we see any need in it.
+
+### Option 2: Extending Runtime Dimensions Support
+
+This option suggests extending the existing runtime dimensions support for
+relevant primitives.
```

Comment:
I don't see option 1 and option 2 are exclusive. Can we have both?

Reply:
You are correct, they are not. It is possible to have a kernel cache while enabling the runtime dimensions feature for specific cases. The RFC is about proposing a general direction to invest in and the kernel cache seems to be a good way to get better performance overall.

Reply:
The major concern of option 1 is the overhead of creating pd and other metadata. Currently, we cache the primitives directly in the frameworks to avoid that. So with options, having primitives supporting variable shape sizes would be beneficial here.

Reply:
> I don't think putting option 1 vs option 2 is the discussion to have. Mostly the discussion should be:
> 
> * what strategy do we need for enabling dynamic shapes? (user guided vs transparent)
> * what levers do we have
> * under which condition each of those levers are efficient.

I believe I didn't make it clear in the RFC that the intention is not to pick the **right** option of the proposed ones. The goal is to define the problem that we have and see what options are available to tackle it, and suggest what we do next.
The suggestion is to:
* Implement an implicit kernel cache to get the better performance overall and **flexibility** to experiment with kernel generalization to increase the kernel cache hit rate
* Enable runtime dimensions, if it's possible to do that without introducing a regression, for optimizing critical cases. For example, the cases when primitive and primitive descriptor creation overhead becomes critical for the workload. This is mostly suggested as an exception to address critical bottlenecks

All the topics from your list above are covered by the RFC.

> * could you add some data on how useful it can be, e.g. with model runs performance data.

We need to keep in mind that the performance will largely depend on whether or not the kernels can be re-used for the given workload. If they cannot be re-used then there will be no performance improvement. As far as I know we are currently targeting a workload that uses matmul primitive whose kernels can be re-used so we should see performance improvement. @rjoursler is collaborating with the corresponding team to get performance data for his implementation of the kernel cache.

> * if needed what changes in primitive implementations themselves would be needed to make that kernel cache hit rate useful. For example, what should be the target hit rate for all primitive for the kernel cache to help with overall performance ? (on GPU, I would expect that if a single primitive for each inference run get recreated everytime, this will kill overall performance, right?). Performance measurement on an example would be immensely valuable here.

I don't think we can define a targe like that. Ideally, each primitive should have a limited, reasonable number of kernels that can be re-used. But in practice, it would be really hard to achieve that so if we the kernels cannot be re-used then neither runtime dimensions nor kernel cache will help here. As for the data, again, it's very workload dependent as a very limited number of primitives can re-use kernels at this point.  This is one the reasons why I suggest to focus on the implicit kernel cache. This way we can experiment with the implementations to see if it's feasible to make the kernels re-usable without hurting the performance. And if we can't achieve that then we will have to accept the limitations of the JIT approach that we have chosen in the past or explore ways to trade performance performance for reusability.

> * a little more about constraints on primitive creation (other than kernel creation). For example primitive "configuration" time should remain reasonable, as kernel cache does not remove it. Again some data with breakdown on some model could be useful.

Right, I mentioned that in one of the cons of the kernel cache option. We have already gone through this when we were working on the primitive cache. Below is a breakdown for ResNet-50 convolutions. Those are not BRGEMM based convolutions but it can give us a gist of what we have for most of the primitives (note, for GPU the kernel generation time  would be multiple times longer). For BRGEMM based primitives the primitive descriptor overhead could be higher due to block calculations, which is a required step to get the best performance. The only way to eliminate that overhead would be adding runtime dimensions support, which would be fine if we could do that and the overhead was critical. Please keep in mind that now we don't have anything to mitigate the creation overhead problem for the cases when workloads use dynamic shapes. With the kernel cache we could remove the kernel generation overhead which is the most expensive part during primitive creation, and the users won't have to even know about it.

![image](https://user-images.githubusercontent.com/11770112/236115185-b83a9a52-81ba-4642-a721-a94912cbe551.png)

Reply:
> The major concern of option 1 is the overhead of creating pd and other metadata. Currently, we cache the primitives directly in the frameworks to avoid that. 

You are right, but the cache doesn't work when the shapes keep changing therefore you have to create a new pd and primitive every for each new set of shapes. The option 2 doesn't eliminate all overhead but it can eliminate the part of it that is usually the most expensive. And given the set problems that comes with the option 1 it sounds reasonable to start with the implicit kernel cache and see how it goes.